### PR TITLE
Fix check for Keeper system path modification

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
keeper-fix: correctly compare paths in write requests to Keeper internal system node paths.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Common/ZooKeeper/ZooKeeperCommon.cpp [hunk 1]
```diff
@@ -898,4 +898,19 @@ ZooKeeperRequestFactory::ZooKeeperRequestFactory()
...
+    auto [first_it, second_it] = std::mismatch(path.begin(), path.end(), match_to.begin(), match_to.end());
+
+    if (second_it != match_to.end())
+        return NOT_MATCH;
+
+    if (first_it == path.end())
+        return EXACT;
+
+    return *first_it == '/' ? IS_CHILD : NOT_MATCH;
+}
```
Comment (yakov-olkhovskiy) [inline src/Common/ZooKeeper/ZooKeeperCommon.cpp:914]: this function seems to be not tolerable to trailing slash on match_to argument - consider:
```
matchPath("/path/file", "/path/");
```
will return `NOT_MATCH`, while it's `IS_CHILD`

**Review by yakov-olkhovskiy — APPROVED** [review]
since `keeper_system_path` doesn't contain trailing slash this seems to be fine, but consider my note on `matchPath` function

File: src/Common/ZooKeeper/ZooKeeperCommon.cpp (hunk 1 above)
Comment (antonio2368) [reply src/Common/ZooKeeper/ZooKeeperCommon.cpp]: nice catch, I handled those cases + added a unit test for some combination of paths.

File: src/Common/ZooKeeper/ZooKeeperCommon.cpp [hunk 2]
```diff
@@ -898,4 +898,31 @@ ZooKeeperRequestFactory::ZooKeeperRequestFactory()
     registerZooKeeperRequest<OpNum::FilteredList, ZooKeeperFilteredListRequest>(*this);
 }
 
+PathMatchResult matchPath(std::string_view path, std::string_view match_to)
+{
+    using enum PathMatchResult;
+
+    if (match_to == "/")
+        return path == "/" ? EXACT : IS_CHILD;
```
Comment (yakov-olkhovskiy) [inline src/Common/ZooKeeper/ZooKeeperCommon.cpp:906]: do we really need this? I guess everything should work without it...

**Review by yakov-olkhovskiy — APPROVED** [review]

File: src/Common/ZooKeeper/ZooKeeperCommon.cpp [hunk 3]
```diff
@@ -898,4 +898,31 @@ ZooKeeperRequestFactory::ZooKeeperRequestFactory()
...
+
+    const auto clean_path = [](auto & p)
+    {
+        if (p.ends_with('/'))
+            p.remove_suffix(1);
+    };
+
+    clean_path(path);
+    clean_path(match_to);
```
Comment (yakov-olkhovskiy) [inline src/Common/ZooKeeper/ZooKeeperCommon.cpp:915]: Also maybe much shorter without such complications:
```suggestion
   if (path.ends_with('/'))
        path.remove_suffix(1);
   if (match_to.ends_with('/'))
        match_to.remove_suffix(1);
```

**Pull request merged**